File: 1114-binary-search-tree-to-greater-sum-tree/1114-binary-search-tree-to-greater-sum-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstToGst(root *TreeNode) *TreeNode {
    arr := [] int{}
    inOrderTraversal(&arr , root)
    for i := len(arr) -2 ;i>= 0;i--{
        arr[i] = arr[i]+ arr[i+1]
    }
   index := 0
   inOrderTraversal2(arr, &index, root)
    return root
}

func inOrderTraversal(arr *[]int , root *TreeNode){
    if root == nil{
        return
    }
    inOrderTraversal(arr, root.Left)
    *arr = append(*arr, root.Val)
    inOrderTraversal(arr, root.Right)
}

func inOrderTraversal2(arr []int, index *int, root *TreeNode) {
    if root == nil {
        return
    }
    inOrderTraversal2(arr, index, root.Left)
    root.Val = arr[*index]
    *index++
    inOrderTraversal2(arr, index, root.Right)
}

File: 1114-binary-search-tree-to-greater-sum-tree/1114-binary-search-tree-to-greater-sum-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstToGst(TreeNode root) {
        /*
         ArrayList<Integer> arr = new ArrayList<>();
        
      
        inOrderTraversal(arr, root);
        
       
        for(int i = arr.size() - 2; i >= 0; i--) {
            arr.set(i, arr.get(i) + arr.get(i + 1));
        }
        
        
        inOrderTraversal2(arr, new int[]{0}, root);
        
        return root; 
        */
        int sum[] = new int[1];
       solve(root , sum);
       return root;
    }
      void solve(TreeNode root , int sum[]){
        if(root == null){
            return ;
        }
        solve(root.right , sum);
        sum[0]  += root.val;
        root.val = sum[0];
         solve(root.left , sum);
      }
    void inOrderTraversal(ArrayList<Integer> arr, TreeNode root) {
        if (root == null) {
            return;
        }
        inOrderTraversal(arr, root.left);
        arr.add(root.val);
        inOrderTraversal(arr, root.right);
    }
    
    void inOrderTraversal2(ArrayList<Integer> arr, int[] index, TreeNode root) {
        if (root == null) {
            return;
        }
        inOrderTraversal2(arr, index, root.left);
        root.val = arr.get(index[0]);
        index[0]++;
        inOrderTraversal2(arr, index, root.right);
    
    }
}